Lots of build errors on Clang/osx:

/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:496:18: error: 'FlushingJsonBuffer::setScalarInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:473:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:497:18: error: 'FlushingJsonBuffer::setScalarUInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:474:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarUInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size);
                 ^
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/hthor.cpp:18:
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:29:
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.hpp:29:
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:496:18: error: 'FlushingJsonBuffer::setScalarInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:473:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:497:18: error: 'FlushingJsonBuffer::setScalarUInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:474:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarUInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size);
                 ^
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/hthorkey.cpp:17:
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:29:
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.hpp:29:
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:496:18: error: 'FlushingJsonBuffer::setScalarInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:473:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:497:18: error: 'FlushingJsonBuffer::setScalarUInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:474:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarUInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size);
                 ^
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.cpp:30:
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.ipp:28:
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.hpp:29:
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:496:18: error: 'FlushingJsonBuffer::setScalarInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:473:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:497:18: error: 'FlushingJsonBuffer::setScalarUInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:474:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarUInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size);
                 ^
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclgraph.cpp:27:
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.ipp:28:
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.hpp:29:
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:496:18: error: 'FlushingJsonBuffer::setScalarInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:473:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:497:18: error: 'FlushingJsonBuffer::setScalarUInt' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiehelper.hpp:474:18: note: hidden overloaded virtual function 'FlushingStringBuffer::setScalarUInt' declared here: different number of parameters (4 vs 6)
    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size);
                 ^
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclgraph.cpp:27:
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.ipp:32:
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:326:18: error: 'CBaseServerDebugContext::addBreakpoint' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void addBreakpoint(IXmlWriter *output, const char *modeString, const char *id, const char *action, 
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:284:18: note: hidden overloaded virtual function 'CBaseDebugContext::addBreakpoint' declared here: different number of parameters (1 vs 10)
    virtual void addBreakpoint(IBreakpointInfo &bp);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:329:18: error: 'CBaseServerDebugContext::removeBreakpoint' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void removeBreakpoint(IXmlWriter *output, unsigned removeIdx);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:285:18: note: hidden overloaded virtual function 'CBaseDebugContext::removeBreakpoint' declared here: different number of parameters (1 vs 2)
    virtual void removeBreakpoint(IBreakpointInfo &bp);
                 ^
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.cpp:30:
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.ipp:32:
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:326:18: error: 'CBaseServerDebugContext::addBreakpoint' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void addBreakpoint(IXmlWriter *output, const char *modeString, const char *id, const char *action, 
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:284:18: note: hidden overloaded virtual function 'CBaseDebugContext::addBreakpoint' declared here: different number of parameters (1 vs 10)
    virtual void addBreakpoint(IBreakpointInfo &bp);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:329:18: error: 'CBaseServerDebugContext::removeBreakpoint' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void removeBreakpoint(IXmlWriter *output, unsigned removeIdx);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:285:18: note: hidden overloaded virtual function 'CBaseDebugContext::removeBreakpoint' declared here: different number of parameters (1 vs 2)
    virtual void removeBreakpoint(IBreakpointInfo &bp);
                 ^
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclgraph.cpp:30:
In file included from /Users/rchapman/HPCC-Platform/common/environment/dalienv.hpp:27:
/Users/rchapman/HPCC-Platform/common/environment/environment.hpp:107:28: error: 'IConstDropZoneServerInfo::getName' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual StringBuffer & getName(StringBuffer & name) const = 0;
                           ^
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.cpp:34:
In file included from /Users/rchapman/HPCC-Platform/common/environment/dalienv.hpp:27:
/Users/rchapman/HPCC-Platform/common/environment/environment.hpp:107:28: error: 'IConstDropZoneServerInfo::getName' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual StringBuffer & getName(StringBuffer & name) const = 0;
                           ^
/Users/rchapman/HPCC-Platform/common/environment/environment.hpp:38/Users/rchapman/HPCC-Platform/common/environment/environment.hpp::2638:: 26: note: notehidden overloaded virtual function 'IConstEnvBase::getName' declared here: type mismatch at 1st parameter ('IStringVal &' vs 'StringBuffer &'): 
hidden overloaded virtual function 'IConstEnvBase::getName' declared here: type mismatch at 1st parameter ('IStringVal &' vs 'StringBuffer &')
    virtual IStringVal & getName(IStringVal & str) const = 0;
                         ^    virtual IStringVal & getName(IStringVal & str) const = 0;

                         ^
/Users/rchapman/HPCC-Platform/common/environment/environment.hpp:143:25:/Users/rchapman/HPCC-Platform/common/environment/environment.hpp :143:error25: : 'IConstDaFileSrvInfo::getName' hides overloaded virtual function [-Werror,-Woverloaded-virtual]error
: 'IConstDaFileSrvInfo::getName' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual const char *getName() const = 0;
                        ^
    virtual const char *getName() const = 0;
                        ^
/Users/rchapman/HPCC-Platform/common/environment/environment.hpp:38:26/Users/rchapman/HPCC-Platform/common/environment/environment.hpp:: 38:note26: : hidden overloaded virtual function 'IConstEnvBase::getName' declared here: different number of parameters (1 vs 0)
note: hidden overloaded virtual function 'IConstEnvBase::getName' declared here: different number of parameters (1 vs 0)
    virtual IStringVal & getName(IStringVal & str) const = 0;
                         ^
    virtual IStringVal & getName(IStringVal & str) const = 0;
                         ^
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/hthor.cpp:27:
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:326:18: error: 'CBaseServerDebugContext::addBreakpoint' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void addBreakpoint(IXmlWriter *output, const char *modeString, const char *id, const char *action, 
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:284:18: note: hidden overloaded virtual function 'CBaseDebugContext::addBreakpoint' declared here: different number of parameters (1 vs 10)
    virtual void addBreakpoint(IBreakpointInfo &bp);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:329:18: error: 'CBaseServerDebugContext::removeBreakpoint' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void removeBreakpoint(IXmlWriter *output, unsigned removeIdx);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:285:18: note: hidden overloaded virtual function 'CBaseDebugContext::removeBreakpoint' declared here: different number of parameters (1 vs 2)
    virtual void removeBreakpoint(IBreakpointInfo &bp);
                 ^
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/hthorkey.cpp:26:
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:326:18: error: 'CBaseServerDebugContext::addBreakpoint' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void addBreakpoint(IXmlWriter *output, const char *modeString, const char *id, const char *action, 
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:284:18: note: hidden overloaded virtual function 'CBaseDebugContext::addBreakpoint' declared here: different number of parameters (1 vs 10)
    virtual void addBreakpoint(IBreakpointInfo &bp);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:329:18: error: 'CBaseServerDebugContext::removeBreakpoint' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    virtual void removeBreakpoint(IXmlWriter *output, unsigned removeIdx);
                 ^
/Users/rchapman/HPCC-Platform/common/thorhelper/roxiedebug.hpp:285:18: note: hidden overloaded virtual function 'CBaseDebugContext::removeBreakpoint' declared here: different number of parameters (1 vs 2)
    virtual void removeBreakpoint(IBreakpointInfo &bp);
                 ^
2 errors generated.
make[3]: *** [ecl/hthor/CMakeFiles/hthorlib.dir/hthorstep.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclgraph.cpp:27:
/Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.ipp:932:23: error: private field 'owner' is not used [-Werror,-Wunused-private-field]
        EclSubGraph  *owner;
                      ^
7 errors generated.
make[3]: *** [ecl/hthor/CMakeFiles/hthorlib.dir/__/eclagent/eclgraph.cpp.o] Error 1
/Users/rchapman/HPCC-Platform/ecl/hthor/hthorkey.cpp:3077:14: error: private field 'subno' is not used [-Werror,-Wunused-private-field]
    unsigned subno;
             ^
5 errors generated.
make[3]: *** [ecl/hthor/CMakeFiles/hthorlib.dir/hthorkey.cpp.o] Error 1
In file included from /Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.cpp:30:
/Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.ipp:932:23: error: private field 'owner' is not used [-Werror,-Wunused-private-field]
        EclSubGraph  *owner;
                      ^
7 errors generated.
make[3]: *** [ecl/hthor/CMakeFiles/hthorlib.dir/__/eclagent/eclagent.cpp.o] Error 1
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/hthor.cpp:18:
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:611:14: error: private field 'activityId' is not used [-Werror,-Wunused-private-field]
    unsigned activityId;
             ^
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:637:10: error: private field 'isVariable' is not used [-Werror,-Wunused-private-field]
    bool isVariable;
         ^
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:921:29: error: private field 'helper' is not used [-Werror,-Wunused-private-field]
    IHThorHashAggregateArg &helper;
                            ^
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:1055:23: error: private field 'helper' is not used [-Werror,-Wunused-private-field]
    IHThorDegroupArg &helper;
                      ^
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:1634:29: error: private field 'helper' is not used [-Werror,-Wunused-private-field]
    IHThorDatasetResultArg &helper;
                            ^
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:1645:25: error: private field 'helper' is not used [-Werror,-Wunused-private-field]
    IHThorRowResultArg &helper;
                        ^
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:1672:16: error: private field 'helper' is not used [-Werror,-Wunused-private-field]
    IHThorArg &helper;
               ^
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:1825:23: error: private field 'helper' is not used [-Werror,-Wunused-private-field]
    IHThorRegroupArg &helper;
                      ^
/Users/rchapman/HPCC-Platform/ecl/hthor/./hthor.ipp:2730:36: error: private field 'helper' is not used [-Werror,-Wunused-private-field]
    IHThorGraphLoopResultWriteArg &helper;
                                   ^
In file included from /Users/rchapman/HPCC-Platform/ecl/hthor/hthor.cpp:52:
/Users/rchapman/HPCC-Platform/ecl/eclagent/eclagent.ipp:932:23: error: private field 'owner' is not used [-Werror,-Wunused-private-field]
        EclSubGraph  *owner;
                      ^

--- a/common/environment/environment.hpp
+++ b/common/environment/environment.hpp
@@ -104,6 +104,7 @@ interface IConstInstanceInfo : extends IConstEnvBase
 
 interface IConstDropZoneServerInfo : extends IConstEnvBase
 {
+    using IConstEnvBase::getName;     // The following function has a different prototype to the base interface
     virtual StringBuffer & getName(StringBuffer & name) const = 0;
     virtual StringBuffer & getServer(StringBuffer & server) const = 0;
 };
@@ -140,6 +141,7 @@ interface IConstDfuQueueInfoIterator : extends IIteratorOf<IConstDfuQueueInfo>
 
 interface IConstDaFileSrvInfo : extends IConstEnvBase
 {
+    using IConstEnvBase::getName;     // The following function has a different prototype to the base interface
     virtual const char *getName() const = 0;
     virtual unsigned getPort() const = 0;
     virtual bool getSecure() const = 0;
@@ -152,6 +154,7 @@ interface IConstInstanceInfoIterator : extends IIteratorOf<IConstInstanceInfo>
 
 interface IConstSparkThorInfo : extends IConstEnvBase
 {
+    using IConstEnvBase::getName;     // The following function has a different prototype to the base interface
     virtual IStringVal & getName(IStringVal & str) const = 0;
     virtual IStringVal & getBuild(IStringVal & str) const = 0;
     virtual IStringVal & getThorClusterName(IStringVal & str) const = 0;

--- a/common/thorhelper/roxiedebug.hpp
+++ b/common/thorhelper/roxiedebug.hpp
@@ -318,40 +318,45 @@ protected:
                                    unsigned hitCount=0, const char *hitCountMode=NULL);
     virtual void waitForDebugger(DebugState state, IActivityDebugContext *probe);
 public:
-    IMPLEMENT_IINTERFACE;
+    IMPLEMENT_IINTERFACE_USING(CBaseDebugContext);
     CBaseServerDebugContext(const IContextLogger &_logctx, IPropertyTree *_queryXGMML) ;
+
+//Avoid complaints about mismatched prototypes for virtuals
+    using CBaseDebugContext::addBreakpoint;
+    using CBaseDebugContext::removeBreakpoint;
+
     void serializeBreakpoints(MemoryBuffer &to);
-    virtual void debugInitialize(const char *id, const char *_queryName, bool _breakAtStart);
-    virtual void debugTerminate();
+    virtual void debugInitialize(const char *id, const char *_queryName, bool _breakAtStart) override;
+    virtual void debugTerminate() override;
     virtual void addBreakpoint(IXmlWriter *output, const char *modeString, const char *id, const char *action, 
                                const char *fieldName, const char *condition, const char *value, bool caseSensitive,
-                               unsigned hitCount, const char *hitCountMode);
-    virtual void removeBreakpoint(IXmlWriter *output, unsigned removeIdx);
-    virtual void removeAllBreakpoints(IXmlWriter *output);
-    virtual void listBreakpoint(IXmlWriter *output, unsigned listIdx) const;
-    virtual void listAllBreakpoints(IXmlWriter *output) const;
-    virtual void debugInterrupt(IXmlWriter *output);
-    virtual void debugContinue(IXmlWriter *output, const char *modeString, const char *id);
-    virtual void debugRun(IXmlWriter *output) ;
-    virtual void debugQuit(IXmlWriter *output) ;
-    virtual void debugSkip(IXmlWriter *output);
-    virtual void debugStatus(IXmlWriter *output) const;
-    virtual void debugStep(IXmlWriter *output, const char *modeString);
-    virtual void debugNext(IXmlWriter *output);
-    virtual void debugOver(IXmlWriter *output);
-    virtual void debugChanges(IXmlWriter *output, unsigned sinceSequence) const;
-    virtual void debugCounts(IXmlWriter *output, unsigned sinceSequence, bool reset);
-    virtual void debugWhere(IXmlWriter *output) const;
-    virtual void debugSearch(IXmlWriter *output, const char *fieldName, const char *condition, const char *value, bool caseSensitive, bool fullRows) const;
-    virtual void debugPrint(IXmlWriter *output, const char *edgeId, unsigned startRow, unsigned numRows) const;
-    virtual void debugGetConfig(IXmlWriter *output, const char *name, const char *id) const;
-    virtual void debugSetConfig(IXmlWriter *output, const char *name, const char *value, const char *id);
-    virtual void getCurrentGraphXGMML(IXmlWriter *output, bool original) const;
-    virtual void getQueryXGMML(IXmlWriter *output) const;
-    virtual void getGraphXGMML(IXmlWriter *output, const char *graphName) const;
-    virtual const char *queryQueryName() const;
-    virtual const char *queryDebugId() const;
-    virtual void debugPrintVariable(IXmlWriter *output, const char *name, const char *type) const;
+                               unsigned hitCount, const char *hitCountMode) override;
+    virtual void removeBreakpoint(IXmlWriter *output, unsigned removeIdx) override;
+    virtual void removeAllBreakpoints(IXmlWriter *output) override;
+    virtual void listBreakpoint(IXmlWriter *output, unsigned listIdx) const override;
+    virtual void listAllBreakpoints(IXmlWriter *output) const override;
+    virtual void debugInterrupt(IXmlWriter *output) override;
+    virtual void debugContinue(IXmlWriter *output, const char *modeString, const char *id) override;
+    virtual void debugRun(IXmlWriter *output)  override;
+    virtual void debugQuit(IXmlWriter *output)  override;
+    virtual void debugSkip(IXmlWriter *output) override;
+    virtual void debugStatus(IXmlWriter *output) const override;
+    virtual void debugStep(IXmlWriter *output, const char *modeString) override;
+    virtual void debugNext(IXmlWriter *output) override;
+    virtual void debugOver(IXmlWriter *output) override;
+    virtual void debugChanges(IXmlWriter *output, unsigned sinceSequence) const override;
+    virtual void debugCounts(IXmlWriter *output, unsigned sinceSequence, bool reset) override;
+    virtual void debugWhere(IXmlWriter *output) const override;
+    virtual void debugSearch(IXmlWriter *output, const char *fieldName, const char *condition, const char *value, bool caseSensitive, bool fullRows) const override;
+    virtual void debugPrint(IXmlWriter *output, const char *edgeId, unsigned startRow, unsigned numRows) const override;
+    virtual void debugGetConfig(IXmlWriter *output, const char *name, const char *id) const override;
+    virtual void debugSetConfig(IXmlWriter *output, const char *name, const char *value, const char *id) override;
+    virtual void getCurrentGraphXGMML(IXmlWriter *output, bool original) const override;
+    virtual void getQueryXGMML(IXmlWriter *output) const override;
+    virtual void getGraphXGMML(IXmlWriter *output, const char *graphName) const override;
+    virtual const char *queryQueryName() const override;
+    virtual const char *queryDebugId() const override;
+    virtual void debugPrintVariable(IXmlWriter *output, const char *name, const char *type) const override;
 };
 
 //=======================================================================================

--- a/common/thorhelper/roxiehelper.cpp
+++ b/common/thorhelper/roxiehelper.cpp
@@ -2370,18 +2370,18 @@ void FlushingJsonBuffer::startScalar(const char *resultName, unsigned sequence, 
     }
 }
 
-void FlushingJsonBuffer::setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size, bool simpleTag, const char *simpleName)
+void FlushingJsonBuffer::setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size)
 {
-    startScalar(resultName, sequence, simpleTag, simpleName);
+    startScalar(resultName, sequence, false, nullptr);
     if (size < 7) //JavaScript only supports 53 significant bits
         s.append(value);
     else
         s.append('"').append(value).append('"');
 }
 
-void FlushingJsonBuffer::setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size, bool simpleTag, const char *simpleName)
+void FlushingJsonBuffer::setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size)
 {
-    startScalar(resultName, sequence, simpleTag, simpleName);
+    startScalar(resultName, sequence, false, nullptr);
     if (size < 7) //JavaScript doesn't support unsigned, and only supports 53 significant bits
         s.append(value);
     else

--- a/common/thorhelper/roxiehelper.hpp
+++ b/common/thorhelper/roxiehelper.hpp
@@ -488,14 +488,14 @@ public:
     {
     }
 
-    void append(double data);
-    void encodeString(const char *x, unsigned len, bool utf8=false);
-    void encodeData(const void *data, unsigned len);
-    void startDataset(const char *elementName, const char *resultName, unsigned sequence, bool _extend = false, const IProperties *xmlns=NULL, bool adaptive=false);
-    void startScalar(const char *resultName, unsigned sequence, bool simpleTag, const char *simplename=nullptr);
-    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
-    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size, bool simpleTag = false, const char *simplename=nullptr);
-    virtual void flushXML(StringBuffer &current, bool isClosing)
+    virtual void append(double data) override;
+    virtual void encodeString(const char *x, unsigned len, bool utf8=false) override;
+    virtual void encodeData(const void *data, unsigned len) override;
+    virtual void startDataset(const char *elementName, const char *resultName, unsigned sequence, bool _extend = false, const IProperties *xmlns=NULL, bool adaptive=false) override;
+    virtual void startScalar(const char *resultName, unsigned sequence, bool simpleTag, const char *simplename=nullptr) override;
+    virtual void setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size) override;
+    virtual void setScalarUInt(const char *resultName, unsigned sequence, unsigned __int64 value, unsigned size) override;
+    virtual void flushXML(StringBuffer &current, bool isClosing) override
     {
         FlushingStringBuffer::flushXML(current, isClosing, (extend) ? "," : nullptr);
     }

--- a/ecl/eclagent/CMakeLists.txt
+++ b/ecl/eclagent/CMakeLists.txt
@@ -39,6 +39,10 @@ include_directories (
          ${CMAKE_BINARY_DIR}/oss
     )
 
+if (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_CLANG)
+  SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wno-switch -Wno-unused-parameter -Werror -Wno-error=delete-non-virtual-dtor")
+endif()
+
 ADD_DEFINITIONS( -DNO_SYBASE -D_CONSOLE )
 
 HPCC_ADD_EXECUTABLE ( hthor ${SRCS} )

--- a/ecl/eclagent/eclagent.cpp
+++ b/ecl/eclagent/eclagent.cpp
@@ -85,7 +85,6 @@ typedef IEclProcess* (* EclProcessFactory)();
 
 constexpr LogMsgCategory MCsetresult = MCprogress(100);     // Category used to inform when setting result
 constexpr LogMsgCategory MCgetresult = MCprogress(200);     // Category used to inform when getting result
-constexpr LogMsgCategory MCresolve = MCprogress(100);       // Category used to inform during name resolution
 constexpr LogMsgCategory MCrunlock = MCprogress(100);      // Category used to inform about run lock progress
 
 Owned<IPropertyTree> agentTopology;
@@ -336,14 +335,11 @@ public:
     virtual void threadmain() override
     {
         StringBuffer rawText;
-        unsigned priority = (unsigned) -2;
-        unsigned memused = 0;
         IpAddress peer;
         bool continuationNeeded;
         bool isStatus;
 
         Owned<IDebuggerContext> debuggerContext;
-        unsigned slavesReplyLen = 0;
         HttpHelper httpHelper(NULL);
         try
         {
@@ -373,7 +369,6 @@ public:
             return;
         }
 
-        bool isRaw = false;
         bool isHTTP = false;
         Owned<IPropertyTree> queryXml;
         StringBuffer sanitizedText;
@@ -397,8 +392,6 @@ public:
             }
             bool isRequest = false;
             bool isRequestArray = false;
-            bool isBlind = false;
-            bool isDebug = false;
 
             sanitizeQuery(queryXml, queryName, sanitizedText, isHTTP, uid, isRequest, isRequestArray);
             DBGLOG("Received debug query %s", sanitizedText.str());
@@ -919,7 +912,6 @@ UChar *EclAgent::getResultVarUnicode(const char * stepname, unsigned sequence)
     PROTECTED_GETRESULT(stepname, sequence, "VarUnicode", "unicode",
         MemoryBuffer result;
         r->getResultUnicode(MemoryBuffer2IDataVal(result));
-        unsigned tlen = result.length()/2;
         result.append((UChar)0);
         return (UChar *)result.detach();
     );
@@ -1437,7 +1429,6 @@ ILocalOrDistributedFile *EclAgent::resolveLFN(const char *fname, const char *err
 
 bool EclAgent::fileExists(const char *name)
 {
-    unsigned __int64 size = 0;
     StringBuffer lfn;
     expandLogicalName(lfn, name);
 
@@ -1619,7 +1610,7 @@ void EclAgent::restoreCluster(IWorkUnit *wu)
 
 unsigned EclAgent::getNodes()//retrieve node count for current cluster
 {
-    if (clusterWidth == -1)
+    if (clusterWidth == (unsigned)-1)
     {
         if (!isStandAloneExe)
         {
@@ -3066,8 +3057,8 @@ char * EclAgent::queryIndexMetaData(char const * lfn, char const * xpath)
         try
         {
             OwnedIFile file = createIFile(part->getFilename(rfn, copy));
-            unsigned __int64 thissize = file->size();
-            if(thissize != -1)
+            offset_t thissize = file->size();
+            if(thissize != (offset_t)-1)
             {
                 StringBuffer remotePath;
                 rfn.getPath(remotePath);
@@ -4014,7 +4005,7 @@ class DebugProbe : public InputProbe, implements IActivityDebugContext
 public:
     DebugProbe(IHThorInput *_in, IEngineRowStream *_stream, unsigned _sourceId, unsigned _sourceIdx, DebugActivityRecord *_sourceAct, unsigned _targetId, unsigned _targetIdx, DebugActivityRecord *_targetAct, unsigned _iteration, unsigned _channel, IDebuggableContext *_debugContext)
         : InputProbe(_in, _stream, _sourceId, _sourceIdx, _targetId, _targetIdx, _iteration, _channel),
-          sourceAct(_sourceAct), targetAct(_targetAct), debugContext(_debugContext)
+          debugContext(_debugContext), sourceAct(_sourceAct), targetAct(_targetAct)
     {
         historyCapacity = debugContext->getDefaultHistoryCapacity();
         nextHistorySlot = 0;

--- a/ecl/eclagent/eclagent.ipp
+++ b/ecl/eclagent/eclagent.ipp
@@ -929,7 +929,6 @@ private:
     class LegacyInputProbe : public CInterface, implements IHThorInput, implements IEngineRowStream
     {
         IHThorInput  *in;
-        EclSubGraph  *owner;
         size32_t    maxRowSize;
         unsigned sourceId;
         unsigned outputIndex;
@@ -939,8 +938,8 @@ private:
     public:
         IMPLEMENT_IINTERFACE;
 
-        LegacyInputProbe(IHThorInput *_in, EclSubGraph *_owner, unsigned _sourceId, int outputidx)
-            : in(_in), owner(_owner), sourceId(_sourceId), outputIndex(outputidx)
+        LegacyInputProbe(IHThorInput *_in, unsigned _sourceId, int outputidx)
+            : in(_in), sourceId(_sourceId), outputIndex(outputidx)
         {
             StringAttrBuilder edgeIdText(edgeId);
             edgeIdText.append(_sourceId).append("_").append(outputidx);
@@ -1010,7 +1009,7 @@ private:
     public:
         virtual IEclGraphResults * resolveLocalQuery(__int64 activityId)
         {
-            if (activityId == container->queryId())
+            if ((unsigned __int64)activityId == container->queryId())
                 return container;
             return ctx->resolveLocalQuery(activityId);
         }
@@ -1050,7 +1049,7 @@ public:
     {
         if (probeEnabled)
         {
-            LegacyInputProbe *probe = new LegacyInputProbe(in, this, sourceId, outputidx);
+            LegacyInputProbe *probe = new LegacyInputProbe(in, sourceId, outputidx);
             probes.append(*probe);
             return probe;
         }

--- a/ecl/eclagent/eclgraph.cpp
+++ b/ecl/eclagent/eclgraph.cpp
@@ -418,7 +418,7 @@ bool EclGraphElement::alreadyUpToDate(IAgentContext & agent)
     }
 
     IPropertyTree & cur = f->queryAttributes();
-    if ((eclCRC != cur.getPropInt("@eclCRC")) || (totalCRC != cur.getPropInt64("@totalCRC")))
+    if ((eclCRC != (unsigned)cur.getPropInt("@eclCRC")) || (totalCRC != (unsigned __int64)cur.getPropInt64("@totalCRC")))
         return false;
     return true;
 }
@@ -754,7 +754,7 @@ IHThorException * EclGraphElement::makeWrappedException(IException * e)
 //---------------------------------------------------------------------------
 
 EclSubGraph::EclSubGraph(IAgentContext & _agent, EclGraph & _parent, EclSubGraph * _owner, unsigned _seqNo, bool enableProbe, CHThorDebugContext * _debugContext, IProbeManager * _probeManager)
-    : parent(_parent), owner(_owner), seqNo(_seqNo), probeEnabled(enableProbe), debugContext(_debugContext), probeManager(_probeManager), isLoopBody(false)
+    : probeEnabled(enableProbe), seqNo(_seqNo), parent(_parent), owner(_owner), debugContext(_debugContext), probeManager(_probeManager), isLoopBody(false)
 {
     executed = false;
     created = false;
@@ -1249,7 +1249,6 @@ void EclGraph::execute(const byte * parentExtract)
     try
     {
         unsigned startTime = msTick();
-        aindex_t lastSink = -1;
         ForEachItemIn(idx, graphs)
         {
             EclSubGraph & cur = graphs.item(idx);

--- a/ecl/hthor/CMakeLists.txt
+++ b/ecl/hthor/CMakeLists.txt
@@ -73,6 +73,10 @@ include_directories (
          ${HPCC_SOURCE_DIR}/common/wuanalysis
     )
 
+if (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_CLANG)
+  SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wno-switch -Wno-unused-parameter -Werror -Wno-error=delete-non-virtual-dtor")
+endif()
+
 ADD_DEFINITIONS( -D_USRDLL -DHTHOR_EXPORTS -DSTARTQUERY_EXPORTS )
 
 HPCC_ADD_LIBRARY( hthorlib SHARED ${SRCS} ${INCLUDES} )

--- a/ecl/hthor/hthor.cpp
+++ b/ecl/hthor/hthor.cpp
@@ -60,8 +60,6 @@
 
 static unsigned const hthorReadBufferSize = 0x10000;
 static offset_t const defaultHThorDiskWriteSizeLimit = I64C(10*1024*1024*1024); //10 GB, per Nigel
-static size32_t const spillStreamBufferSize = 0x10000;
-static unsigned const hthorPipeWaitTimeout = 100; //100ms - fairly arbitrary choice
 
 using roxiemem::IRowManager;
 using roxiemem::OwnedRoxieRow;
@@ -193,11 +191,9 @@ bool isRemoteReadCandidate(const IAgentContext &agent, const RemoteFilename &rfn
 
 //=====================================================================================================
 
-CHThorActivityBase::CHThorActivityBase(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorArg & _help, ThorActivityKind _kind, EclGraph & _graph) : agent(_agent), help(_help),  outputMeta(help.queryOutputMeta()), kind(_kind), activityId(_activityId), subgraphId(_subgraphId), graph(_graph)
+CHThorActivityBase::CHThorActivityBase(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorArg & _help, ThorActivityKind _kind, EclGraph & _graph)
+: help(_help), kind(_kind), graph(_graph), agent(_agent), outputMeta(help.queryOutputMeta()), activityId(_activityId), subgraphId(_subgraphId)
 {
-    input = NULL;
-    processed = 0;
-    rowAllocator = NULL;    
 }
 
 void CHThorActivityBase::setInput(unsigned index, IHThorInput *_input)
@@ -337,7 +333,7 @@ class CHThorClusterWriteHandler : public ClusterWriteHandler
     IAgentContext &agent;
 public:
     CHThorClusterWriteHandler(char const * _logicalName, char const * _activityType, IAgentContext &_agent) 
-        : agent(_agent), ClusterWriteHandler(_logicalName, _activityType)
+        : ClusterWriteHandler(_logicalName, _activityType), agent(_agent)
     {
     }
 
@@ -2810,7 +2806,7 @@ bool HashDedupTable::insertBest(const void * nextrow)
 }
 
 CHThorHashDedupActivity::CHThorHashDedupActivity(IAgentContext & _agent, unsigned _activityId, unsigned _subgraphId, IHThorHashDedupArg & _arg, ThorActivityKind _kind, EclGraph & _graph)
-: CHThorSimpleActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg), table(_arg, activityId), hashTableFilled(false), hashDedupTableIter(table)
+: CHThorSimpleActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg), table(_arg), hashTableFilled(false), hashDedupTableIter(table)
 {
     keepBest = helper.keepBest();
 }
@@ -3458,9 +3454,9 @@ const void * CHThorAggregateActivity::nextRow()
 //=====================================================================================================
 
 CHThorHashAggregateActivity::CHThorHashAggregateActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorHashAggregateArg &_arg, ThorActivityKind _kind, EclGraph & _graph, bool _isGroupedAggregate)
-: CHThorSimpleActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg),
-  isGroupedAggregate(_isGroupedAggregate),
-  aggregated(_arg, _arg)
+: CHThorSimpleActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph),
+  aggregated(_arg, _arg),
+  isGroupedAggregate(_isGroupedAggregate)
 {
 }
 
@@ -3853,7 +3849,7 @@ bool CHThorChooseSetsEnthActivity::includeRow(const void * row)
 
 //=====================================================================================================
 
-CHThorDegroupActivity::CHThorDegroupActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorDegroupArg &_arg, ThorActivityKind _kind, EclGraph & _graph) : CHThorSteppableActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg)
+CHThorDegroupActivity::CHThorDegroupActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorDegroupArg &_arg, ThorActivityKind _kind, EclGraph & _graph) : CHThorSteppableActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph)
 {
 }
 
@@ -5598,7 +5594,8 @@ const void * CHThorLookupJoinActivity::LookupTable::doFind(const void * left) co
     return NULL;
 }
 
-CHThorLookupJoinActivity::CHThorLookupJoinActivity(IAgentContext & _agent, unsigned _activityId, unsigned _subgraphId, IHThorHashJoinArg &_arg, ThorActivityKind _kind, EclGraph & _graph) : CHThorActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg), table(0), outBuilder(NULL)
+CHThorLookupJoinActivity::CHThorLookupJoinActivity(IAgentContext & _agent, unsigned _activityId, unsigned _subgraphId, IHThorHashJoinArg &_arg, ThorActivityKind _kind, EclGraph & _graph)
+ : CHThorActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg), outBuilder(NULL), table(0)
 {
 }
 
@@ -6530,7 +6527,7 @@ const void *CHThorInlineTableActivity::nextRow()
 
 //=====================================================================================================
 
-CHThorNullActivity::CHThorNullActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorArg &_arg, ThorActivityKind _kind, EclGraph & _graph) : CHThorSimpleActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg)
+CHThorNullActivity::CHThorNullActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorArg &_arg, ThorActivityKind _kind, EclGraph & _graph) : CHThorSimpleActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph)
 {
 }
 
@@ -6770,7 +6767,7 @@ const void *CHThorNonEmptyActivity::nextRow()
 
 //=====================================================================================================
 
-CHThorRegroupActivity::CHThorRegroupActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorRegroupArg &_arg, ThorActivityKind _kind, EclGraph & _graph) : CHThorMultiInputActivity(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg)
+CHThorRegroupActivity::CHThorRegroupActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorRegroupArg &_arg, ThorActivityKind _kind, EclGraph & _graph) : CHThorMultiInputActivity(_agent, _activityId, _subgraphId, _arg, _kind, _graph)
 {
 }
 
@@ -7349,7 +7346,7 @@ void CHThorTopNActivity::ready()
 {
     CHThorSimpleActivityBase::ready();
     limit = helper.getLimit();
-    assertex(limit == (size_t)limit);
+    assertex(limit == (__int64)(size_t)limit);
     sorted = (const void * *)checked_calloc((size_t)(limit+1), sizeof(void *), "topn");
     sortedCount = 0;
     curIndex = 0;
@@ -7809,7 +7806,7 @@ void CHThorResultActivity::extractResult(unsigned & retSize, void * & ret)
 //=====================================================================================================
 
 CHThorDatasetResultActivity::CHThorDatasetResultActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorDatasetResultArg &_arg, ThorActivityKind _kind, EclGraph & _graph)
- : CHThorResultActivity(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg)
+ : CHThorResultActivity(_agent, _activityId, _subgraphId, _arg, _kind, _graph)
 {
 }
 
@@ -7834,7 +7831,7 @@ void CHThorDatasetResultActivity::execute()
 //=====================================================================================================
 
 CHThorRowResultActivity::CHThorRowResultActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorRowResultArg &_arg, ThorActivityKind _kind, EclGraph & _graph)
- : CHThorResultActivity(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg)
+ : CHThorResultActivity(_agent, _activityId, _subgraphId, _arg, _kind, _graph)
 {
 }
 
@@ -8170,7 +8167,6 @@ void CHThorDiskReadBaseActivity::ready()
 
     unsigned expectedCrc = helper.getDiskFormatCrc();
     unsigned projectedCrc = helper.getProjectedFormatCrc();
-    unsigned actualCrc = expectedCrc;
     IDistributedFile *dFile = nullptr;
     if (ldFile)
         dFile = ldFile->queryDistributedFile();  // Null for local file usage
@@ -8434,7 +8430,6 @@ bool CHThorDiskReadBaseActivity::openNext()
               (!dfsParts&&(partNum<ldFile->numParts())))
         {
             IDistributedFilePart * curPart = dfsParts?&dfsParts->query():NULL;
-            IDistributedFile *dFile = ldFile->queryDistributedFile();  // Null for local file usage
 
             unsigned numCopies = curPart?curPart->numCopies():ldFile->numPartCopies(partNum);
             //MORE: Order of copies should be optimized at this point....
@@ -9938,7 +9933,7 @@ const void *CHThorGraphLoopResultReadActivity::nextRow()
 //=====================================================================================================
 
 CHThorGraphLoopResultWriteActivity::CHThorGraphLoopResultWriteActivity(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorGraphLoopResultWriteArg &_arg, ThorActivityKind _kind, EclGraph & _graph, __int64 graphId)
- : CHThorActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph), helper(_arg)
+ : CHThorActivityBase(_agent, _activityId, _subgraphId, _arg, _kind, _graph)
 {
     graph = resolveLocalQuery(graphId);
 }
@@ -11203,7 +11198,6 @@ const void *CHThorNewDiskReadActivity::nextRow()
                 const byte * next = (const byte *)inputRowStream->nextRow(nextSize);
                 if (!isSpecialRow(next))
                 {
-                    size32_t thisSize = 0;
                     if (likely(!hasMatchFilter || helper.canMatch(next)))
                     {
                         size32_t thisSize = helper.transform(outBuilder.ensureRow(), next);

--- a/ecl/hthor/hthor.ipp
+++ b/ecl/hthor/hthor.ipp
@@ -200,21 +200,22 @@ class CHThorActivityBase : public CInterface, implements IHThorActivity, impleme
 {
 protected:
     enum ActivityState { StateCreated, StateReady, StateDone };
-    IHThorInput *input;
+    IHThorInput *input = nullptr;
     IHThorArg & help;
     ThorActivityKind kind;
     EclGraph & graph;
-    CHThorActivityBase(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorArg & _help, ThorActivityKind _kind, EclGraph & _graph);
-    ~CHThorActivityBase();
     IAgentContext &agent;
     CachedOutputMetaData outputMeta;
-    unsigned __int64 processed;
-    unsigned __int64 initialProcessed;
+    unsigned __int64 processed = 0;
+    unsigned __int64 initialProcessed = 0;
     unsigned activityId;
     unsigned subgraphId;
+    IEngineRowAllocator *rowAllocator = nullptr;
+
+    CHThorActivityBase(IAgentContext &_agent, unsigned _activityId, unsigned _subgraphId, IHThorArg & _help, ThorActivityKind _kind, EclGraph & _graph);
+    ~CHThorActivityBase();
     IException * makeWrappedException(IException * e) const;
     IException * makeWrappedException(IException * e, char const * extra) const;
-    IEngineRowAllocator *rowAllocator;
 
 public:
     IMPLEMENT_IINTERFACE;
@@ -567,9 +568,8 @@ private:
 class HashDedupTable : public SuperHashTable
 {
 public:
-    HashDedupTable(IHThorHashDedupArg & _helper, unsigned _activityId)
-        : helper(_helper), 
-          activityId(_activityId)
+    HashDedupTable(IHThorHashDedupArg & _helper)
+        : helper(_helper)
     {
         queryBestCompare = helper.queryCompareBest();
     }
@@ -607,7 +607,6 @@ public:
 
 private:
     IHThorHashDedupArg & helper;
-    unsigned activityId;
     Owned<IEngineRowAllocator> keyRowAllocator;
     ICompare * queryBestCompare;
 };
@@ -633,7 +632,6 @@ class CHThorNormalizeActivity : public CHThorSimpleActivityBase
 {
     IHThorNormalizeArg &helper;
     OwnedConstRoxieRow inbuff;
-    bool isVariable;
     unsigned numThisRow;
     unsigned curRow;
     unsigned __int64 numProcessedLastGroup;
@@ -917,11 +915,10 @@ public:
 
 class CHThorHashAggregateActivity : public CHThorSimpleActivityBase
 {
-    IHThorHashAggregateArg &helper;
     RowAggregator aggregated;
 
-    bool eof;
-    bool gathered;
+    bool eof = false;
+    bool gathered = false;
     bool isGroupedAggregate;
 public:
     CHThorHashAggregateActivity(IAgentContext &agent, unsigned _activityId, unsigned _subgraphId, IHThorHashAggregateArg &_arg, ThorActivityKind _kind, EclGraph & _graph, bool _isGroupedAggregate);
@@ -951,7 +948,7 @@ public:
 class CHThorFirstNActivity : public CHThorSimpleActivityBase
 {
     IHThorFirstNArg &helper;
-    __int64 doneThisGroup = 0;
+    __uint64 doneThisGroup = 0;
     __uint64 limit = 0;  // You would think int was enough for most practical cases...
     __uint64 skip = 0;
     bool finished = false;
@@ -1051,7 +1048,6 @@ public:
 
 class CHThorDegroupActivity : public CHThorSteppableActivityBase
 {
-    IHThorDegroupArg &helper;
 public:
     CHThorDegroupActivity(IAgentContext &agent, unsigned _activityId, unsigned _subgraphId, IHThorDegroupArg &_arg, ThorActivityKind _kind, EclGraph & _graph);
 
@@ -1118,8 +1114,8 @@ protected:
 class CSimpleSorterBase : public ISorter, public CInterface
 {
 public:
-    CSimpleSorterBase(ICompare * _compare, roxiemem::IRowManager * _rowManager, size32_t _initialSize, size32_t _commitDelta) : compare(_compare), finger(0), rowManager(_rowManager),
-        rowsToSort(_rowManager, _initialSize, _commitDelta, UNKNOWN_ROWSET_ID) {}
+    CSimpleSorterBase(ICompare * _compare, roxiemem::IRowManager * _rowManager, size32_t _initialSize, size32_t _commitDelta) : rowManager(_rowManager), compare(_compare),
+        rowsToSort(_rowManager, _initialSize, _commitDelta, UNKNOWN_ROWSET_ID), finger(0) {}
     virtual ~CSimpleSorterBase()                            { killSorted(); }
     IMPLEMENT_IINTERFACE;
     virtual bool addRow(const void * next)                  { return rowsToSort.append(next); }
@@ -1144,7 +1140,7 @@ public:
 
 protected:
     roxiemem::IRowManager * rowManager;
-    unsigned activityId;
+    unsigned activityId = 0;
     ICompare * compare;
     DynamicRoxieOutputRowArray rowsToSort;
     aindex_t finger;
@@ -1167,7 +1163,8 @@ public:
 class CStableSorter : public CSimpleSorterBase
 {
 public:
-    CStableSorter(ICompare * _compare, roxiemem::IRowManager * _rowManager, size32_t _initialSize, size32_t _commitDelta, roxiemem::IBufferedRowCallback * _rowCB) : CSimpleSorterBase(_compare, _rowManager, _initialSize, _commitDelta), commitDelta(_commitDelta), index(NULL), indexCapacity(0) {}
+    CStableSorter(ICompare * _compare, roxiemem::IRowManager * _rowManager, size32_t _initialSize, size32_t _commitDelta, roxiemem::IBufferedRowCallback * _rowCB)
+     : CSimpleSorterBase(_compare, _rowManager, _initialSize, _commitDelta), index(NULL), indexCapacity(0), commitDelta(_commitDelta) {}
     virtual ~CStableSorter() { killSorted(); }
 
     virtual bool addRow(const void * next);
@@ -1629,8 +1626,6 @@ public:
 
 class CHThorDatasetResultActivity : public CHThorResultActivity
 {
-    IHThorDatasetResultArg &helper;
-
 public:
     IMPLEMENT_SINKACTIVITY;
 
@@ -1640,8 +1635,6 @@ public:
 
 class CHThorRowResultActivity : public CHThorResultActivity
 {
-    IHThorRowResultArg &helper;
-
 public:
     IMPLEMENT_SINKACTIVITY;
 
@@ -1667,7 +1660,6 @@ public:
 
 class CHThorNullActivity : public CHThorSimpleActivityBase
 {
-    IHThorArg &helper;
 public:
     CHThorNullActivity(IAgentContext &agent, unsigned _activityId, unsigned _subgraphId, IHThorArg &_arg, ThorActivityKind _kind, EclGraph & _graph);
 
@@ -1820,7 +1812,6 @@ public:
 
 class CHThorRegroupActivity : public CHThorMultiInputActivity
 {
-    IHThorRegroupArg &helper;
     unsigned inputIndex;
     bool eof;
     unsigned __int64 numProcessedLastGroup;
@@ -2725,7 +2716,6 @@ public:
 
 class CHThorGraphLoopResultWriteActivity : public CHThorActivityBase
 {
-    IHThorGraphLoopResultWriteArg &helper;
     ILocalEclGraphResults * graph;
 
 public:

--- a/ecl/hthor/hthorkey.cpp
+++ b/ecl/hthor/hthorkey.cpp
@@ -47,8 +47,8 @@ static IKeyIndex *openKeyFile(IDistributedFilePart & keyFile)
         try
         {
             OwnedIFile ifile = createIFile(keyFile.getFilename(rfn,copy));
-            unsigned __int64 thissize = ifile->size();
-            if (thissize != -1)
+            offset_t thissize = ifile->size();
+            if (thissize != (offset_t)-1)
             {
                 StringBuffer remotePath;
                 rfn.getPath(remotePath);
@@ -74,19 +74,6 @@ static IKeyIndex *openKeyFile(IDistributedFilePart & keyFile)
     keyFile.getFilename(rfn).getRemotePath(url);
     throw MakeStringException(1001, "Could not open key file at %s%s", url.str(), (numCopies > 1) ? " or any alternate location." : ".");
 }
-
-static void setProgress(IPropertyTree &node, const char *name, const char *value)
-{
-    StringBuffer attr("@");
-    node.setProp(attr.append(name).str(), value);
-}
-
-static void setProgress(IPropertyTree &node, const char *name, unsigned __int64 value)
-{
-    StringBuffer attr("@");
-    node.setPropInt64(attr.append(name).str(), value);
-}
-
 
 class TransformCallback : public CInterface, implements IThorIndexCallback 
 {
@@ -1707,20 +1694,19 @@ class ThreadedPartHandler : public CInterface
 {
 protected:
     Linked<IThreadPool> threadPool;
-    PooledThreadHandle threadHandle;
+    PooledThreadHandle threadHandle = 0;
     QueueOf<ROW, true> pending;
     CriticalSection crit;
     Semaphore limit;
-    bool started;
+    bool started = false;
     Owned<IDistributedFilePart> part;
     IThreadedExceptionHandler *handler;
 
 public:
     typedef ThreadedPartHandler<ROW> SELF;
     ThreadedPartHandler(IDistributedFilePart *_part, IThreadedExceptionHandler *_handler, IThreadPool * _threadPool)
-        : limit(MAX_FETCH_LOOKAHEAD), part(_part), handler(_handler), threadHandle(0), threadPool(_threadPool)
+        : threadPool(_threadPool), limit(MAX_FETCH_LOOKAHEAD), part(_part), handler(_handler)
     {
-        started = false;
     }
 
     ~ThreadedPartHandler()
@@ -1833,8 +1819,8 @@ public:
           encryptionkey(_encryptionkey), 
           activityId(_activityId), 
           outputMeta(_outputMeta),
-          prefetcher(_prefetcher),
-          rowAllocator(_rowAllocator)
+          rowAllocator(_rowAllocator),
+          prefetcher(_prefetcher)
     {
         base = _base;
         top = _base + _size;
@@ -1871,8 +1857,8 @@ public:
             try
             {
                 OwnedIFile ifile = createIFile(part->getFilename(rfn,copy));
-                unsigned __int64 thissize = ifile->size();
-                if (thissize != -1)
+                offset_t thissize = ifile->size();
+                if (thissize != (offset_t)-1)
                 {
                     IPropertyTree & props = part->queryAttributes();
                     unsigned __int64 expectedSize;
@@ -1885,7 +1871,7 @@ public:
                         expectedSize = props.getPropInt64("@compressedSize", -1);
                     else
                         expectedSize = props.getPropInt64("@size", -1);
-                    if(thissize != expectedSize && expectedSize != -1)
+                    if(thissize != expectedSize && expectedSize != (unsigned __int64)-1)
                         throw MakeStringException(0, "File size mismatch: file %s was supposed to be %" I64F "d bytes but appears to be %" I64F "d bytes", ifile->queryFilename(), expectedSize, thissize); 
                     if(blockcompressed)
                         rawFile.setown(createCompressedFileReader(ifile,eexp));
@@ -1980,7 +1966,7 @@ protected:
     static offset_t getPartSize(IDistributedFilePart *part)
     {
         offset_t partsize = part->queryAttributes().getPropInt64("@size", -1);
-        if (partsize==-1)
+        if (partsize == (offset_t)-1)
         {
             MTIME_SECTION(queryActiveTimer(), "Fetch remote file size");
             unsigned numCopies = part->numCopies();
@@ -1991,7 +1977,7 @@ protected:
                 {
                     OwnedIFile ifile = createIFile(part->getFilename(rfn,copy));
                     partsize = ifile->size();
-                    if (partsize != -1)
+                    if (partsize != (offset_t)-1)
                     {
                         // TODO: Create DistributedFilePropertyLock for parts
                         part->lockProperties();
@@ -2007,7 +1993,7 @@ protected:
                 }
             }
         }
-        if (partsize==-1)
+        if (partsize == (offset_t)-1)
             throw MakeStringException(0, "Unable to determine size of filepart"); 
         return partsize;
     }
@@ -2892,7 +2878,7 @@ public:
 
     IMPLEMENT_IINTERFACE;
 
-    CJoinGroup(const void *_left, IJoinProcessor *_join, CJoinGroup *_groupStart) : join(_join), matches(*this)
+    CJoinGroup(const void *_left, IJoinProcessor *_join, CJoinGroup *_groupStart) : matches(*this),join(_join)
     {
         candidates = 0;
         left = _left;
@@ -2981,7 +2967,7 @@ protected:
     unsigned matchcount;
     CIArrayOf<MatchSet> matchsets;
     atomic_t endMarkersPending;
-    IJoinProcessor *join;
+    IJoinProcessor *join = nullptr;
     mutable CriticalSection crit;
     CJoinGroup *groupStart;
     unsigned candidates;
@@ -3088,7 +3074,6 @@ class KeyedLookupPartHandler : extends ThreadedPartHandler<MatchSet>, implements
     Owned<IKeyManager> manager;
     IAgentContext &agent;
     DistributedKeyLookupHandler * tlk;
-    unsigned subno;
 
 public:
     IMPLEMENT_IINTERFACE;
@@ -3164,7 +3149,7 @@ public:
     IMPLEMENT_IINTERFACE;
 
     DistributedKeyLookupHandler(IDistributedFile *f, IJoinProcessor &_owner, IAgentContext &_agent)
-        : file(f), owner(_owner), agent(_agent)
+        : owner(_owner), file(f), agent(_agent)
     {
         threadPool.setown(createThreadPool("hthor keyed join lookup thread pool", &threadFactory));
         IDistributedSuperFile *super = f->querySuperFile();
@@ -3259,7 +3244,7 @@ public:
 };
 
 KeyedLookupPartHandler::KeyedLookupPartHandler(IJoinProcessor &_owner, IDistributedFilePart *_part, DistributedKeyLookupHandler * _tlk, unsigned _subno, IThreadPool * _threadPool, IAgentContext &_agent)
-    : owner(_owner), ThreadedPartHandler<MatchSet>(_part, _tlk, _threadPool), agent(_agent), tlk(_tlk), subno(_subno)
+    : ThreadedPartHandler<MatchSet>(_part, _tlk, _threadPool), owner(_owner), agent(_agent), tlk(_tlk)
 {
 }
 

--- a/plugins/cassandra/CMakeLists.txt
+++ b/plugins/cassandra/CMakeLists.txt
@@ -70,6 +70,13 @@ if(USE_CASSANDRA)
       set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-copy")
     endif ()
   endif()
+  if (CMAKE_COMPILER_IS_CLANG)
+  # From time to time need to add flags in here to disable warnings, if a new compiler picks up something new.
+  # e.g.
+    if (NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 10.0.0)
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-copy")
+    endif ()
+  endif()
   if (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_CLANG)
     remove_definitions (-fsanitize=undefined -fno-sanitize=alignment -fsanitize-trap=undefined)
   endif()

--- a/thorlcr/msort/tsortm.cpp
+++ b/thorlcr/msort/tsortm.cpp
@@ -1175,11 +1175,13 @@ public:
                             splitMap.setown(CalcPartitionUsingSampling());
                         else
 #endif
+                        {
 #ifdef TRACE_PARTITION
                             splitMap.setown(CalcPartition(true));
 #else
                             splitMap.setown(CalcPartition(false));
 #endif
+                        }
                     }
                     if (!partitioninfo->splitkeys.checkSorted(icompare))
                     {

